nvaccess/nvda
# winVersion/Windows 10 version fetch: use Windows Registry instead of a static release names to builds

Hi,

Mostly in preparation for a possibility that Windows 10 Version 21H2 will be used by multiple builds, including Server 2022, and prepares NVDA to support Windows 11 and future Windows versions:

### Link to issue number:
Closes #12509 

### Summary of the issue:
Currently a static Windows 10 release names to builds map is used to record Windows 10 builds to feature updates (releases). Until 21H1 there was a one-to-one correspondence between releases and builds, but it appears Version 21H2 will be represented by multiple builds:

* Windows 10: 19044
* Server 2022: 20348
* Windows 11: build 22000

Builds subject to confirmation.

Therefore this breaks current static map.

### Description of how this pull request fixes the issue:
Since Version 1511, Windows Registry holds release name which corresponds to builds (e.g. 1511 = 10586, 20H2 = 19042), housed inside a "Release Id" key. Starting with Version 20H2, a separate "DisplayVersion" key is used to announce public release name. In Windows Insider Preview builds, dev channel has display version of "Dev", making it even easier to determine what kind of build a user is running. This tradition continues with Windows 11 and Server 2022, with Registry recording "21H2", which makes sense given their projected release period. Therefore use Windows Registry instead of the static releases to builds map and deprecate releases to builds map.

### Testing strategy:
Manual testing:

1. On supported Windows 10 releases (1507 LTS, 1607/server 2016 LTS, 1809/Server 2019 LTS, 1909 (Enterprise and Education), 2004, 20H2, 21H1, 21H2/Server 2022 preview, Windows Insider Preview dev channel), open Windows Registry path described in #12509 and confirm that release Id's and display name keys are present (the latter is available starting with 20H2).
2. In Python Console, verify that winVersion.getWinVer() says "Windows 10 YYMM/YYHn" and the last part (release name) matches the value obtained from Windows Registry.
3. Optionally, define a local unit test that checks the version string returned by winVersion.getWinVer() starts with "Windows 10 release" for the local system (e.g. "Windows 10 21H1" on build 19043).

### Known issues with pull request:
None

### Change log entries:
Likely a change for developers, stating that on Windows 10, release names are obtained from Windows Registry, although note that this is an internal operation. A visible changelog is:

Changes for developers:

On Windows 10 Version 1511 and later (including Insider Preview builds), current Windows feature update release name is obtained from Windows Registry. (#12509)
The releases to builds mapping found in winVersion module is deprecated and will be removed in 2022.1. (#12544)

### Code Review Checklist:



  - Speech 
  - Braille
  - Low Vision
  - Different web browsers

### Discussion:
If this PR is adopted, what should we do about the static release names to builds map? There are three possibilities:

1. Swap releases with builds so the map records build : release. This makes Windows Registry method unnecessary - adding future Windows releases will continue to involve editing this map.
2. Deprecate and remove the static map - the only place the map is used is when fetching Windows 10 release names anyway, and since Windows Registry contains needed information, there is no need for this map. If deprecation and removal are chosen, a follow-up PR should be created during 2022.1 development cycle to remove the static map altogether.
3. For releases represented by multiple builds, have Windows 10 release name fetcher check for these builds and return the release name for these. This makes version fetch routine complicated as one needs to hunt for all build numbers represented by a single release name.

Update (2021-07-07): deprecated and replaced by a private builds to releases mapping.

Thanks.
